### PR TITLE
Ashdaniels25

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/kata1.fizzBuzz.js [hunk 1]
```diff
@@ -1,3 +1,14 @@
...
+  if (number % 3 === 0 && number % 5 === 0) {
+    return "FizzBuzz";
+  }
+  if (number % 3 === 0) {
+    return "Fizz";
+  }
+  if (number % 5 === 0) {
+    return "Buzz";
+  }
+  return number;
```
Comment (PerryBaran) [inline src/kata1.fizzBuzz.js:11]: You can use variables to store the results of `number % 3 === 0` and `number % 5 === 0` to make your code more DRY. i.e.
```
const fizzBuzz = (number) => {
  const isDivisableBy3 = number % 3 === 0;
  const isDivisableBy5 = number % 5 === 0;

  if (isDivisableBy3 && isDivisableBy5) {
    return "FizzBuzz";
  }
  if (isDivisableBy5) {
    return "Buzz";
  }
  if (isDivisableBy3) {
    return "Fizz";
  }

  return number;
};
```

File: test/kata1.fizzBuzz.test.js [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
+    expect(fizzBuzz(5)).toBe("Buzz");
+    expect(fizzBuzz(10)).toBe("Buzz");
+    expect(fizzBuzz(20)).toBe("Buzz");
+  });
 
-  test("returns FizzBuzz when passed a multiple 3 and 5", () => {});
-
-  test("returns the number when it isn't a multiple of 3 or 5", () => {});
+  test("returns 'FizzBuzz' when passed a multiple of both 3 and 5", () => {
+    // Test with a multiple of both 3 and 5
```
Comment (PerryBaran) [inline test/kata1.fizzBuzz.test.js:19]: Do not need these comments here, as what the test is testing for is already mentioned in the test description

File: src/kata2.booleanToWord.js [hunk 3]
```diff
@@ -1,3 +1,3 @@
-const booleanToWord = (boolean) => {};
+const booleanToWord = (boolean) => (boolean ? "Yes" : "No");
```
Comment (PerryBaran) [inline src/kata2.booleanToWord.js:1]: Good

File: test/kata2.booleanToWord.test.js [hunk 4]
```diff
@@ -1,5 +1,13 @@
...
-});
+  test("converts true to 'Yes'", () => {
+    const result = booleanToWord(true);
+    expect(result).toBe("Yes");
+  });
+
+  test("converts false to 'No'", () => {
+    const result = booleanToWord(false);
+    expect(result).toBe("No");
+  });
```
Comment (PerryBaran) [inline test/kata2.booleanToWord.test.js:12]: Good

Comment (PerryBaran) [reply test/kata2.booleanToWord.test.js:12]: There is a linting issue at the bottom of this page (red squiggly line) because it wants an extra line after your final closing bracket

File: src/kata3.numberToReversedDigits.js [hunk 5]
```diff
@@ -1,3 +1,22 @@
-const numberToReversedDigits = (number) => {};
+const numberToReversedDigits = (number) => {
+  if (typeof number !== "number" || isNaN(number)) {
```
Comment (PerryBaran) [inline src/kata3.numberToReversedDigits.js:2]: You have a linting error here as it wants you to use `Number.isNaN(number)` instead of just `isNaN(number)`

File: src/kata3.numberToReversedDigits.js [hunk 6]
```diff
@@ -1,3 +1,22 @@
...
+  let num = Math.abs(number);
+
+  while (num > 0) {
+    const digit = num % 10;
+    digits.push(digit);
+    num = Math.floor(num / 10);
+  }
+
+  return digits;
+};
```
Comment (PerryBaran) [inline src/kata3.numberToReversedDigits.js:20]: Interesting approach. You can also tackle this problem using inbuilt methods. i.e.
```
const numberToReversedDigits = (number) => {
  return number
    .toString()
    .split("")
    .reverse()
    .map((string) => parseInt(string, 10));
}
```

File: test/kata3.numberToReversedDigits.test.js [hunk 7]
```diff
@@ -1,5 +1,17 @@
...
+    const result1 = numberToReversedDigits(12345);
+    expect(result1).toEqual([5, 4, 3, 2, 1]);
+
+    // Test with a single-digit number
+    const result2 = numberToReversedDigits(7);
+    expect(result2).toEqual([7]);
+
+    // Test with a zero
+    const result3 = numberToReversedDigits(0);
+    expect(result3).toEqual([0]);
```
Comment (PerryBaran) [inline test/kata3.numberToReversedDigits.test.js:15]: If you want to have different test scenarios, you should create different tests for each rather then leaving comments about what each is testing. Generally, comments should only be used to explain why you are doing something a certain way, not what you are doing. e.g.
```
describe("numberToReversedDigits", () => {
  test("returns a reversed array of the number's digits", () => {
    const result1 = numberToReversedDigits(12345);
    expect(result1).toEqual([5, 4, 3, 2, 1]);
  });

  test("returns a reversed array for a single digit number", () => {
    const result2 = numberToReversedDigits(7);
    expect(result2).toEqual([7]);
  });

  test("returns a reversed array when the number is 0", () => {
    const result3 = numberToReversedDigits(0);
    expect(result3).toEqual([0]);
  });
});

```

Comment (PerryBaran) [reply test/kata3.numberToReversedDigits.test.js:15]: You've also added functionality that isn't tested here. (the bit where you check if the input is a number or not). You should try to test as much as a function as possible. You can test for a thrown error like:
```
test("throws an error if the input is not a number", () => {
    expect(() => numberToReversedDigits()).toThrow(
      "Input is not a valid number"
    );
  });
```
Note: to test for a thrown error you have to call the function inside of a callback function. This will be covered in more detail later in the track

File: src/kata4.humanCatDogYears.js [hunk 8]
```diff
@@ -1,3 +1,19 @@
...
+  }
+
+  if (number > 2) {
+    const catYears = 24 + (number - 2) * 4;
+    const dogYears = 24 + (number - 2) * 5;
+    return [number * 15, catYears, dogYears];
+  }
+
+  throw new Error("Input is not a valid age");
+};
```
Comment (PerryBaran) [inline src/kata4.humanCatDogYears.js:17]: Your return value is a little wrong here. The first value in the array should just be the humanYears, i.e. `number`. Also, this code doesn't work for `0` years, which is a valid age. To edit your existing function to fix this functionality would look something like:
```
const humanCatDogYears = (number) => {
  if (number === 0) {
    return [number, 0, 0];
  }

  if (number === 1) {
    return [number, 15, 15];
  }

  if (number === 2) {
    return [number, 24, 24];
  }

  if (number > 2) {
    const catYears = 24 + (number - 2) * 4;
    const dogYears = 24 + (number - 2) * 5;
    return [number, catYears, dogYears];
  }

  throw new Error("Input is not a valid age");
};
```

File: test/kata4.humanCatDogYears.test.js [hunk 9]
```diff
@@ -1,3 +1,9 @@
 const { humanCatDogYears } = require("../src");
 
-// Look Ma, no handlebars!!!
+describe("humanCatDogYears", () => {
+  it("calculates cat and dog years based on human years", () => {
+    // Test with a human age of 5 years
+    const result = humanCatDogYears(5);
+    expect(result).toEqual([75, 36, 39]);
+  });
+});
```
Comment (PerryBaran) [inline test/kata4.humanCatDogYears.test.js:9]: You should be adding separate tests for each edge case. i.e. when human years is: less than 0, 0, 1, 2, 3 and more than 3.

File: test/kata4.humanCatDogYears.test.js [hunk 10]
```diff
@@ -1,3 +1,9 @@
 const { humanCatDogYears } = require("../src");
 
-// Look Ma, no handlebars!!!
+describe("humanCatDogYears", () => {
+  it("calculates cat and dog years based on human years", () => {
+    // Test with a human age of 5 years
+    const result = humanCatDogYears(5);
+    expect(result).toEqual([75, 36, 39]);
```
Comment (PerryBaran) [inline test/kata4.humanCatDogYears.test.js:7]: As previously mentioned, the first value in the array should be the input into the function, i.e.
```
    const result = humanCatDogYears(5);
    expect(result).toEqual([5, 36, 39]);
```

File: src/kata5.reachDestination.js [hunk 11]
```diff
@@ -1,3 +1,6 @@
-const reachDestination = (distance, speed) => {};
+const reachDestination = (distance, speed) => {
+  const arrivalTime = Math.ceil((distance / speed) * 2) / 2;
+  return `I will be there in ${arrivalTime}, see you soon.`;
```
Comment (PerryBaran) [inline src/kata5.reachDestination.js:3]: Be there in how long? hours? minutes? years? You should include these units in the string. e.g.
```
 return `I will be there in ${arrivalTime} hours, see you soon.`;
```
However, then if `arrivalTime` is `1` your sentence would then be grammatically incorrect `"I will be there in 1 hours, see you soon."` instead of `"I will be there in 1 hour, see you soon."`. To fix this, you can conditionally add the `"s"` at the end of `"hour"` e.g.
```
return `I will be there in ${arrivalTime} hour${arrivalTime === 1 ? "" : "s"}, see you soon.`;
```

File: test/kata5.reachDestination.test.js [hunk 12]
```diff
@@ -1,5 +1,9 @@
 const { reachDestination } = require("../src");
 
 describe("reachDestination", () => {
-  test("returns string with estimated time of arrival", () => {});
-});
+  test("returns string with estimated time of arrival", () => {
+    expect(reachDestination(44, 10)).toEqual(
+      "I will be there in 4.5, see you soon."
+    );
+  });
```
Comment (PerryBaran) [inline test/kata5.reachDestination.test.js:8]: As previously mentioned, you should add the units for time to this string. After doing this, I would add a separate test for your edge case where the arrival time is `1` e.g.
```
test("returns a string with the estimated time of arrival for 1 hour", () => {
    expect(reachDestination(1, 1)).toEqual("I will be there in 1 hour, see you soon.");
  });
```

File: test/kata6.joinNames.test.js [hunk 13]
```diff
@@ -1,5 +1,8 @@
 const { joinNames } = require("../src");
 
 describe("joinNames", () => {
-  test("returns string of names, seperated by commas and an ampersand", () => {});
+  test("returns string of names, seperated by commas and an ampersand", () => {
+    const inputNames = [{ name: "Bart" }, { name: "Lisa" }, { name: "Maggie" }];
+    expect(joinNames(inputNames)).toEqual("Bart, Lisa & Maggie");
+  });
 });
```
Comment (PerryBaran) [inline test/kata6.joinNames.test.js:8]: Should add more tests for edge cases. I would have separate test for when `inputNames` has a length of: 0, 1, 2, 3 and more than 3.

I have tested for each of these scenarios and your code works as expected. You should still create these test to confirm that and in case someone wants to go back and alter this function later, they can then be sure that the functionality hasn't been broken and get a better understanding of how this function should work with different inputs.
e.g.
```
  test("returns string of name for array of length 1", () => {
    expect(joinNames([{ name: "Bart" }])).toBe("Bart");
  });

  test("returns string names, separated by ampersand for array of length 2", () => {
    expect(joinNames([{ name: "Bart" }, { name: "Lisa" }])).toBe("Bart & Lisa");
  });

  test("returns string of names, separated by commas and an ampersand, for long arrays", () => {
    expect(
      joinNames([
        {
          name: "Bart",
        },
        {
          name: "Lisa",
        },
        {
          name: "Maggie",
        },
        {
          name: "Homer",
        },
        {
          name: "Marge",
        },
      ])
    ).toBe("Bart, Lisa, Maggie, Homer & Marge");
  });

  test("returns an empty string for empty arrays", () => {
    expect(joinNames([])).toBe("");
  });
```

File: src/kata7.getEmployerRole.js [hunk 14]
```diff
@@ -1,3 +1,9 @@
-const getEmployerRole = (employeeName, employees) => {};
+const getEmployerRole = (employeeName, employees) => {
+  for (const employee of employees) {
+    if (employee.name === employeeName) {
+      return employee.role;
+    }
+  }
```
Comment (PerryBaran) [inline src/kata7.getEmployerRole.js:6]: Good solution, you could also do create this functionality using `.find`, e.g.
```
return employees.find(employee => employee.name === employeeName).role;
```

**Review by PerryBaran — COMMENTED** [review]
Overall really good. I've left comments throughout on where you can improve. key points:
- Test as much functionality as you can. Try to think of the edge cases you can test for.
- There is some minor linting issues. If you have a red squiggly line in your code and the file name shows up in red you should try to fix these issues. If you hover your mouse over the red squiggly line it will give you information on what is causing the problem. At the bottom of this box is the text `Quick Fix (Ctrl+.)` (different symbol to `Ctrl` on mac). Clicking this button will bring up another popup with suggestions on how to fix/workaround this problem. The first option in this popup will usually fix your problem for you.
- You should only add comments that explain why you did something the way you did it, not what you are doing.
